### PR TITLE
Milestone 1: Enumerable module with several methods

### PR DESCRIPTION
**Pull request by RaminMammadzada**
In this project we created an Enumerable module. 
The foowing custom enumerable methods are implemented:
```#my_each``` 
```#my_each_with_index``` 
```#my_select``` 
```#my_all?``` 
```#my_any?``` 
```#my_none?``` 
```#my_count``` 
```#my_map```  
```#my_inject```

At the same time we added #muptiply_els which multiplies all the elements of the array together by using #my_inject .

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,154 @@
+# rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity,Metrics/ModuleLength
+module Enumerable
+  def my_each()
+    return self unless block_given?
+
+    length.times do |index|
+      yield(self[index])
+    end
+  end
```
Comment (expjazz) [inline enumerable.rb]: - [x] This should return an Enum if no block is given
image

- [x] It should return self if block is given

- [x] when a block is given and when self is a range, returns the range itself after calling the given block once for each element in self

- [x] when a block is given and when self is a range, it behaves like #each and calls the given block once for each element in self

- [x] when a block is given when self is a hash, it behaves like #each and calls the given block once for each element in self

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+  def my_each_with_index()
+    return self unless block_given?
+
+    length.times do |index|
+      yield(self[index], index)
+    end
+  end
+
```
Comment (expjazz) [inline enumerable.rb]: - [x] This should also return an enumerator if no block is given.

- [x] It also should return self if the block is given and self is an array

- [x] when a block is given and  when self is a range, it should return the range itself after calling the given block once for each element in self

- [x]  when a block is given, and when self is a range, it behaves like #each_with_index calls the given block once for each element in self

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    return self unless block_given?
+
+    Array(self).my_each do |item|
+      next unless yield(item)
+
+      result.push(item)
+    end
+
+    result
+  end
```
Comment (expjazz) [inline enumerable.rb]: - [x] It also needs to return an enumerator if no block is given

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+          next if item.is_a? argument
+        elsif (item.is_a? Numeric) || (item.is_a? String) # When it is a value
+          next if item == argument
+        elsif (argument.class == Regexp) && !(item =~ argument).nil?
+          next
+        end
+      elsif yield(item)
+        next
+      end
+      return false
```
Comment (expjazz) [inline enumerable.rb]: - [x] when a Regex is passed as an argument, it returns true if all of the collection matches the Regex

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+          next unless item == argument
+        elsif (argument.class == Regexp) && (item =~ argument).nil?
+          next
+        end
+      else
+        next unless yield(item)
+      end
+      return true
+    end
+    false
```
Comment (expjazz) [inline enumerable.rb]: - [x] when a Regex is passed as an argument, it should return true if any of the collection matches the Regex

File: enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        elsif (argument.class == Regexp) && (item =~ argument).nil?
+          next
+        end
+      else
+        next unless yield(item)
+      end
+      return false
+    end
+    true
+  end
```
Comment (expjazz) [inline enumerable.rb]: - [x] when a Regex is passed as an argument, it should returns false if any of the collection matches the Regex

File: enumerable.rb [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        next if item != argument
+
+        counter += 1
+      elsif yield(item) == true
+        counter += 1
+      end
+    end
+
+    counter
+  end
```
Comment (expjazz) [inline enumerable.rb]: - [x] It should handle range as well

File: enumerable.rb [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    aux_ary = []
+
+    if argument.class == Proc
+      Array(self).my_each do |item|
+        aux_ary.push(argument.call(item))
+      end
+      return aux_ary
+    end
+
+    return self unless block_given?
```
Comment (expjazz) [inline enumerable.rb]: - [x] It also should return an enumerator if no block is given

File: enumerable.rb [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    end
+    Array(self).my_each_with_index do |item, index|
+      next if skip_flag && index.zero?
+
+      if block_given?
+        acum = yield(acum, item)
+      elsif arguments[0].class == Symbol
+        acum = acum.send(arguments[0], item)
+      elsif arguments[0].is_a? Numeric
+        acum = acum.send(arguments[1], item)
```
Comment (expjazz) [inline enumerable.rb]: - [x] it is raising a "LocalJumpError" when no block or argument is given

**Review by expjazz — CHANGES_REQUESTED** [review]
## Status: Changes Required :recycle: 

Hey Team. You did a Very good job on this project :clap: 

The code is clean, and I can see that you tried to test your code. That is always the correct mentality :1st_place_medal: 

However, there are a few problems (you can check them with detail in the inline comments). Here are the general details:

- [ ] You should handle hashes and ranges.

- [ ] My_each should return self if a block is given, and an enum if no block is given

- [ ] You should handle regex as an argument.

I know this can be tricky, but you really did a great job already :clap: 

### Please make the required changes and resubmit for a new code review once you are ready
Good Luck and Happy Coding :pray:

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Expedito Andrade).:smile:_

**Comment by bertil291utn** [comment]
# **Status:** Changes requested 🟠 
<p align="center"> image <br><strong>Not yet</strong> </p>

# Hey guys, you've been doing a great job, and also I hope you enjoyed coding.

### Your work is amazing you master it GitHub Flow you have a professional README file. 

However, in order to accomplish as the actual page, you required to do these changes, listed below:
- [ ] Please fix this [requirement](https://github.com/RaminMammadzada/ruby-enumerable-methods/pull/1/files#r469498026) 
  - If you still have issue please check this [one](https://github.com/bertil291utn/enumerabe-methods/blob/development/enumerable_method_asignment.rb)

### Summary

-  Check [Jump Error](https://user-images.githubusercontent.com/24902525/90093013-28068880-dcf0-11ea-8fc9-973e025e905c.png) when no block is given. `#my_inject` method



***

Guys work in the requested changes above and request another review.
Good luck. 

> “Never stop, never conform until what is good becomes even better, and what’s better becomes excellent.”

**Comment by RaminMammadzada** [comment]
> # **Status:** Changes requested 🟠
> image
> **Not yet**
> 
> # Hey guys, you've been doing a great job, and also I hope you enjoyed coding.
> ### Your work is amazing you master it GitHub Flow you have a professional README file.
> However, in order to accomplish as the actual page, you required to do these changes, listed below:
> 
> * [ ]  Please fix this [requirement](https://github.com/RaminMammadzada/ruby-enumerable-methods/pull/1/files#r469498026)
>   
>   * If you still have issue please check this [one](https://github.com/bertil291utn/enumerabe-methods/blob/development/enumerable_method_asignment.rb)
> 
> ### Summary
> * Check [Jump Error](https://user-images.githubusercontent.com/24902525/90093013-28068880-dcf0-11ea-8fc9-973e025e905c.png) when no block is given. `#my_inject` method
> 
> Guys work in the requested changes above and request another review.
> Good luck.
> 
> > “Never stop, never conform until what is good becomes even better, and what’s better becomes excellent.”

Please check the previous review. In the last part of expjazz's review he said that  it is raising a "LocalJumpError" when no block or argument is given. Although we that in the original #inject enumerable it gives "LocalJumpError" when NO block or argument is given. So who should we follow ?

**Review by bertil291utn — CHANGES_REQUESTED** [review]
# **Status:** Changes requested 🟠 
<p align="center"> image <br><strong>Not yet</strong> </p>

# Hey guys, you've been doing a great job, and also I hope you enjoyed coding.

### Your work is amazing you master it GitHub Flow you have a professional README file. 

However, in order to accomplish as the actual page, you required to do these changes, listed below:
- [ ] Please fix this [requirement](https://github.com/RaminMammadzada/ruby-enumerable-methods/pull/1/files#r469498026) 
  - If you still have issue please check this [one](https://github.com/bertil291utn/enumerabe-methods/blob/development/enumerable_method_asignment.rb)

### Summary
- Please you have to make changes before sending for a review 
-  Check [Jump Error](https://user-images.githubusercontent.com/24902525/90093013-28068880-dcf0-11ea-8fc9-973e025e905c.png) when no block is given. `#my_inject` method



***

Guys work in the requested changes above and request another review.
Good luck. 

> “Never stop, never conform until what is good becomes even better, and what’s better becomes excellent.”

**Comment by krys2fa** [comment]
# Status: Project Approved! 🥇 
Congratulations @RaminMammadzada and @mateomh! 
You made all the requested changes and got the project approved. 👍 

Your hard work has really paid off. You can now merge this branch into the master branch and go on with your next project.

💡  Do not forget to submit your project completion form after merging this branch.

All the best team,
Chris

**Pull request merged**